### PR TITLE
`Exam mode`: Allow instructors to publish example solutions for exam exercises

### PR DESCRIPTION
**Pull request by akesfeden**
**NOTE: Contains a database migration. It is approved for deploying to test servers.**

#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added `@PreAuthorize` and checked the course groups for all new REST Calls (security).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] Following the [theming guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

Continuation PR of https://github.com/ls1intum/Artemis/pull/4328 and https://github.com/ls1intum/Artemis/pull/4795 to add the feature requested in https://github.com/ls1intum/Artemis/issues/16 for exam exercises. 

This PR adds a new `example_solution_publication_date` field to `Exam` to allow instructors to publish the example solutions to students after the specified publication date. When an example solution is published, students can compare their own submission with the example to realize their mistakes or important details they miss. This can help them to learn the subject better and may improve their points in future exercises and exams. 

### Description

This PR adds a new `example_solution_publication_date` field to `Exam` that serves as a common publication date for all exercises of the exam. It adds `View Example Solution` and `Download Solution Repo` buttons to exam summary view when the example solution publication date passes. `View Example Solution` button navigates the user to a page similar to exercise detail page for students where they can see the problem statement and the example solution. `Download Solution Repo` allows the users to download the solution repo of the programming exercises without the `.git` folder so the students cannot access commit history which can contain sensitive data.

Also, this PR improves the styling of the `View Result`, `Plagiarism Case` and `View Example Solution` buttons in the exam summary view for programming exercises so that the buttons align with the title and problem statement better and they do not push whole explanation to the left of the screen which can be problematic for small screens. See screenshots for a before/after comparison.

The database migration is approved for deploying to test servers.

### Steps for Testing (Includes Exam Mode Testing)

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Exam with a Text, Modeling, File Upload and Programming Exercises

~Main functionality is tested and approved by multiple testers and developers, we need to focus on step 3. and 4. below to make sure exam conduction is working as before.~

1. Log in to Artemis
2. Navigate to Course Administration
3. As a student, participate in the exam and submit a result (You do not have to answer all exercises).
4. Ensure exam conduction is not changed and example solution is not visible to students during the exam (you can check the requests in DevTools to be sure).
5. As an instructor, try setting the example solution publication date before the exam end date and ensure that it fails validation.
6. Then set the example solution publication date to a date between the  exam end date and current time.
7. Ensure you can see the `View Result` and `View Example Solution` buttons in the exam summary view as an instructor and as a student.
8. Ensure you can view or download the examples solutions by clicking these buttons as an instructor and as a student.

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2
#### Exam Mode Test
- [x] Test

#### ~(Client Coverage is given by Codecov Report below)~
| Class/File | Branch | Line |
|------------|-------:|-----:|
| BaseExercise.java | 87% | 98% |
| Exercise.java | 82% | 92% |
| Exam.java | 85% | 93% |
| ExamAccessService.java | 83% | 93% |
| ExamService.java | 83% | 95% |
| ExamResource.java | 91% | 96% |
| ExerciseResource.java | 77% | 95% |
| ProgrammingExerciseExportImportResource.java | 81% | 98% |
| exam.model.ts | 100% | 100% |
| exam-update.component.ts | 87.96% | 96.24% |
| programming-exercise-example-solution-repo-download.component.ts | 100% | 100% |
| example-solution.component.ts | 76.92% | 100% |
| exercise.service.ts | 80.61% | 82.01% |
| course-exercise-details.component.ts | 63.57% | 83.17% |

**Note:** Removed Codecov Report from here since its percentages sometimes differ compared to Bamboo tests. You can check the Codecov Report [here](https://github.com/ls1intum/Artemis/pull/5831#issuecomment-1317673317).

### Screenshots

image

image

image

image

image

#### Button Styling Improvement

**Before**
image

**After**
image

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/domain/BaseExercise.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/domain/BaseExercise.java:180]: Do we even need this abstract method here?

File: src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java:156]:
```suggestion
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java:175]: I'd only filter this if the user is not at least a tutor

File: src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java:155]: Isn't this more or less a normal rest call to retrieve the exercise with an additional check if the example solution is already published?
If this includes the example solution as an attribute in the exercise you might have to check that it is not also included in many other rest calls returning an exercise

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

File: src/main/java/de/tum/in/www1/artemis/domain/BaseExercise.java
Comment (akesfeden) [reply src/main/java/de/tum/in/www1/artemis/domain/BaseExercise.java:180]: The unchanged version had the method declared and defined in `BaseExercise.java` so I kept it that way. After your comment I checked the usages and I couldn't see anything preventing us from moving it to `Exercise.java` so I removed the abstract method declaration.

File: src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java
Comment (akesfeden) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java:155]: This endpoint is intended to fetch the most basic exercise info without any categories, team config, posts, etc. to only get enough data to fill the screen below. I think having a separate endpoint for this has 2 advantages:

- Performance: Number of joins and extra queries are reduced.
- Simplicity & Security: Existing `exercises/{exerciseId}` and `/exercises/:exerciseId/details` are implemented with the assumption of students are not able to access the exam exercises at all, so I thought instead of relaxing the access rules there and making sure we do not leak any sensitive data, a new endpoint that just fetches the required data would be safer and makes the intent clearer.

image

> If this includes the example solution as an attribute in the exercise you might have to check that it is not also included in many other rest calls returning an exercise

Overrides of `filterSensitiveInformation` at TextExercise, ModelingExercise etc. remove the example solution if it is not ready to be published yet from other endpoints that the students are allowed to access.

File: src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java
Comment (akesfeden) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java:175]: Since this endpoint is intended to serve the example solution and the instructors can examine the exam summary from the perspective of a student at `/course-management/:courseId/exams/:examId/student-exams/:studentExamId/summary` at the current state I thought it would be better to filter it for instructors and tutors too so that we do not accidentally rely on a data that is not accessible to the students. Also I think if weneed extra data for the instructors or tutors for a future feature it would be easy to just add an extra condition to skip filtering sensitive data according to user role.

**Review by tobias-lippert — DISMISSED** [review]
Tested in testing session on ts1. Works as expected.

**Review by minrows — APPROVED** [review]
Tested on ts1 during the testing session. Functionality-wise everything works as expected. 👍

**Review by pellumbbaboci — DISMISSED** [review]
Tested during testing session on TS1. Everything works as expected.

**Review by timo07k — APPROVED** [review]
Everything works as expected (tested on imported Final Exam with many exercises including all 4 types). Tested on TS1.

**Review by PhilipWerz — DISMISSED** [review]
Everything worked fine. Created exam with the four expected exercise types and followed all the steps. Tested on ts1

**Review by JohannesStoehr — APPROVED** [review]
Thanks for fixing my comments and clearing things up.
Code looks good to me

**Review by evgeny1984 — DISMISSED** [review]
Tested on ts2 and everything works. Code looks also good to me.

**Comment by evgeny1984** [comment]
Probably some small remark for the UI: no tooltip by the **Example Solution Publication Date**, an icon with explanation would be nice, same as it is displayed on another date time fields.

**Review by evgeny1984 — APPROVED** [review]
Tested on ts2. Works as expected, thank you for the integration of my remarks.

**Review by JohannesStoehr — DISMISSED** [review]
Looked through the new code changes and checked the new translations.

**Review by minrows — APPROVED** [review]

**Review by tobias-lippert — DISMISSED** [review]
Reapprove

**Comment by krusche** [comment]
Please fix the merge conflict before we can merge this PR

**Comment by krusche** [comment]
This PR affects exams, so please make sure to also tick off all tasks in the problem statement description after careful reviewing and testing

**Review by sleiss — APPROVED** [review]
The code looks good me

**Review by TheZoker — APPROVED** [review]
Tested on TS1 in the exam conduction and worked!

**Review by danielbode — APPROVED** [review]
Tested exam conduction on ts1.

**Comment by akesfeden** [comment]
We tested exam conduction in today's testing session on TS1 and did not observe any abnormality or leaked sensitive data so I marked the PR as ready to merge.

**Review by rriyaldhi — APPROVED** [review]
Tested on TS1 (Step 3-4).

**Review by evgeny1984 — APPROVED** [review]
Reapproved as it worked as expected on ts1, especially steps 3,4.

**Review by JohannesStoehr — DISMISSED** [review]
Checked code of latest merges. Looks good to me

**Review by stefanwaldhauser — DISMISSED** [review]
Changes look good to me!

File: src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java
Comment (frederik-kt) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java:150]:
```suggestion
     * @return the ResponseEntity with status 200 (OK) and with the body of the exercise with its example solution after filtering sensitive data, or with
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java
Comment (frederik-kt) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java:151]:
```suggestion
     * status 404 (Not Found) if the exercise is not found, or with status 403 (Forbidden) if the current user does not have access to the example solution.
```

File: src/main/webapp/i18n/de/exam.json
Comment (frederik-kt) [inline src/main/webapp/i18n/de/exam.json:358]:
```suggestion
            "exampleSolutionPublicationDateTooltip": "Studierende können die Beispiellösungen für alle Aufgaben in der Klausur ab diesem Zeitpunkt einsehen, wenn sie an der Klausur teilgenommen haben. Zum Deaktivieren leer lassen.",
```

**Review by frederik-kt — COMMENTED** [review]
Found some typos. Code looks good otherwise.

**Comment by akesfeden** [comment]
Thanks @frederik-kt, I applied your typo fixes.

**Review by pellumbbaboci — DISMISSED** [review]
Tested on ts1 during testing session, works as expected.

**Review by minrows — DISMISSED** [review]
Retested on ts1 during the testing session. LGTM 👍

**Review by rriyaldhi — DISMISSED** [review]
Tested on TS1 including the exam mode as an instructor and works as expected!

**Comment by krusche** [comment]
The coverage of server files is missing. Please add it.
Also the coverage for client files is sometimes below the required 80%.
Is that a mistake in the coverage plugin or is the coverage too low?

**Comment by akesfeden** [comment]
@krusche The Server Test Coverage is written under `Test Coverage` and above `Codecov Report` headers. I got the names of the changed files via `git diff --name-only origin/develop...` and wrote coverages for all of them but if there are any missing files you noticed I can add them too.

I will check the client coverages in Bamboo and add another comment here.

**Comment by akesfeden** [comment]
@krusche I found there are differences between Bamboo report and Codecov report. Codecov reports percentages for some files lower than Bamboo (exam-update.component.ts, example-solution.component.ts, exercise.service.ts, course-exercise-details.component.ts). I updated the description with the coverage percentages from Bamboo.

However I noticed the coverage for `course-exercise-details.component.ts` fell below the expected threshold because of moving already tested parts to another file, I'll increase the coverage for that.

**Review by evgeny1984 — DISMISSED** [review]
Tested on TS2, also the exam mode as well. Everything works as expected. The code changes are also looking good.

**Review by evgeny1984 — DISMISSED** [review]
Re-approved due to the increased coverage changes. No code changes introduced.

**Review by minrows — APPROVED** [review]
Reapproval for updated test coverage. 👍

**Review by rriyaldhi — APPROVED** [review]
Reapproved. Updated code does not change the implementation since last review.

**Review by pellumbbaboci — DISMISSED** [review]
Re-approved after test coverage increase and tested again on ts1.

**Comment by krusche** [comment]
@akesfeden please resolve the merge conflicts in the client tests

**Comment by akesfeden** [comment]
@krusche I resolved the conflict. Bamboo client, server and E2E tests are successful.

**Comment by akesfeden** [comment]
Resolved the merge conflict [here](https://github.com/ls1intum/Artemis/pull/5831/files/189df5490e1e75750037775ba6f55f51843967a2..9423fcc3c660c4c59f2d20e1cc0afafed691e8cc) and tested on TS5 and checked [course-exercise-details.component.html](https://github.com/ls1intum/Artemis/pull/5831/files/189df5490e1e75750037775ba6f55f51843967a2..9423fcc3c660c4c59f2d20e1cc0afafed691e8cc#diff-c3d0ba979f10adfd47112d9bee6db84c92492e9b814ddb08c1f8ccfe88679c7d) still behaving as expected.

**Comment by krusche** [comment]
@akesfeden please resolve the merge conflicts

**Review by stefanwaldhauser — APPROVED** [review]
Tested it in testing session and worked as expected

**Review by TheZoker — APPROVED** [review]
Tested on TS1 during testing session (no code review). Works :)

**Review by pellumbbaboci — APPROVED** [review]
Tested during testing session on TS1, works as expected.

**Review by simon-hng — APPROVED** [review]
Tested on TS1

**Pull request merged**
